File: crates/jsbos/demo-hooks.js
const { 
  Agent, 
  HookEvent, 
  ConfigLoader,
} = require('./index.js');


async function loadConfig() {
  const loader = new ConfigLoader();
  loader.discover();
  const cfg = JSON.parse(loader.loadSync());
  const agentCfg = cfg.agent || {};
  const globalCfg = cfg.global_model || {};
  
  return {
    name: agentCfg.name || 'assistant',
    model: globalCfg.model || 'gpt-4.1',
    apiKey: globalCfg.api_key || '',
    baseUrl: globalCfg.base_url || 'https://api.openai.com/v1',
    systemPrompt: agentCfg.system_prompt || 'You are a helpful assistant.',
    temperature: agentCfg.temperature ?? 0.7,
    timeoutSecs: agentCfg.timeout_secs ?? 120,
  };
}


async function main() {
  console.log('JSBOS Hook & Plugin Full Demo');
  console.log('='.repeat(50));
  
  const config = await loadConfig();
  console.log(`\nModel: ${config.model}`);
  
  const agent = await Agent.create(config);
  
  console.log('\n--- Registering all 7 hook events ---');
  const hookLog = [];
  
  const hookEvents = [
    HookEvent.BeforeLlmCall,
    HookEvent.AfterLlmCall,
    HookEvent.BeforeToolCall,
    HookEvent.AfterToolCall,
    HookEvent.OnMessage,
    HookEvent.OnComplete,
    HookEvent.OnError,
  ];
  
  for (const event of hookEvents) {
    await agent.registerHook(event, async (ctx) => {
      hookLog.push(event);
      console.log(`  [${event}] fired`);
      return 'Continue';
    });
  }
  console.log('All 7 hooks registered');
  
  console.log('\n=== TEST 1: run_simple() ===');
  hookLog.length = 0;
  console.log('\nRunning: agent.runSimple("What is 2+2?")');
  try {
    const result = await agent.runSimple('What is 2+2?');
    console.log(`Result: ${result}`);
  } catch (e) {
    console.log(`Error: ${e.message || e}`);
  }
  console.log(`\nHooks fired: ${hookLog.length}`);
  hookLog.forEach(h => console.log(`  - ${h}`));
  
  console.log('\n=== TEST 2: Error handling ===');
  hookLog.length = 0;
  const badConfig = { ...config, apiKey: 'invalid-key' };
  const badAgent = await Agent.create(badConfig);
  
  for (const event of hookEvents) {
    await badAgent.registerHook(event, async (ctx) => {
      hookLog.push(event);
      console.log(`  [${event}] fired`);
      return 'Continue';
    });
  }
  
  console.log('\nRunning with invalid API key...');
  try {
    const result = await badAgent.runSimple('test');
    console.log(`Result: ${result}`);
  } catch (e) {
    console.log(`Error (expected): ${e.message || e}`);
  }
  console.log(`\nHooks fired: ${hookLog.length}`);
  hookLog.forEach(h => console.log(`  - ${h}`));
  
  console.log('\n' + '='.repeat(50));
  console.log('SUMMARY');
  console.log('='.repeat(50));
  console.log(`
Hook Events:
  - BeforeLlmCall: Before sending request to LLM
  - AfterLlmCall: After receiving response from LLM
  - BeforeToolCall: Before executing a tool
  - AfterToolCall: After tool execution completes
  - OnMessage: When message is added to conversation
  - OnComplete: When agent completes successfully
  - OnError: When an error occurs

Hook Decisions:
  - Continue: Proceed normally
  - Abort: Stop current operation
  - Error(msg): Return error to caller
`);

  badAgent.close();
  agent.close();
}


main()
  .then(() => process.exit(0))
  .catch((err) => {
    console.error(err);
    process.exit(1);
  });

File: crates/pybos/demo_hooks.py
import asyncio
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "brainos"))

from pybos import Agent, AgentConfig, HookEvent, PythonTool


def load_config():
    from pybos import ConfigLoader
    loader = ConfigLoader(strategy="deep_merge")
    loader.discover()
    cfg = loader.load_sync()
    agent_cfg = cfg.get("agent", {})
    global_cfg = cfg.get("global_model", {})
    return AgentConfig(
        name=agent_cfg.get("name", "assistant"),
        model=global_cfg.get("model", "gpt-4.1"),
        api_key=global_cfg.get("api_key", os.getenv("OPENAI_API_KEY", "")),
        base_url=global_cfg.get("base_url", os.getenv("OPENAI_BASE_URL", "https://api.openai.com/v1")),
        system_prompt=agent_cfg.get("system_prompt", "You are a helpful assistant."),
        temperature=agent_cfg.get("temperature", 0.7),
        timeout_secs=agent_cfg.get("timeout_secs", 120),
    )


def make_hooks_demo():
    print("=" * 60)
    print("DEMO: All Hook Events")
    print("=" * 60)
    
    config = load_config()
    agent = Agent.from_config(config)
    
    print("\n1. Adding a tool for react() testing...")
    def add_numbers(a: int, b: int) -> str:
        return str(a + b)
    
    tool = PythonTool(
        name="add",
        description="Add two numbers",
        parameters='{"a": {"type": "int"}, "b": {"type": "int"}}',
        schema='{"a": {"type": "int"}, "b": {"type": "int"}}',
        callback=add_numbers
    )
    agent.add_tool(tool)
    print("   Tool 'add' registered")
    
    print("\n2. Registering all 7 hook events...")
    hook_log = []
    
    def log_hook(event, ctx):
        hook_log.append(str(event))
        return "continue"
    
    for event_name in ["BeforeLlmCall", "AfterLlmCall", "BeforeToolCall", 
                       "AfterToolCall", "OnMessage", "OnComplete", "OnError"]:
        agent.register_hook(HookEvent(event_name), log_hook)
    
    print("   All 7 hooks registered: BeforeLlmCall, AfterLlmCall, BeforeToolCall, AfterToolCall, OnMessage, OnComplete, OnError")
    
    return agent, hook_log


async def test_run_simple(agent, hook_log):
    print("\n" + "=" * 60)
    print("TEST 1: run_simple() - LLM hooks only")
    print("=" * 60)
    
    hook_log.clear()
    print("\nRunning: agent.run_simple('What is 2+2?')")
    try:
        result = await agent.run_simple("What is 2+2?")
        print(f"Result: {result}")
    except Exception as e:
        print(f"Error: {e}")
    
    print(f"\nHooks fired: {len(hook_log)}")
    for h in hook_log:
        print(f"  - {h}")
    
    return hook_log


async def test_react_with_tool(agent, hook_log):
    print("\n" + "=" * 60)
    print("TEST 2: react() with tool call")
    print("=" * 60)
    
    hook_log.clear()
    print("\nRunning: agent.react('What is 5+3? Use the add tool.')")
    try:
        result = await agent.react("What is 5+3? Use the add tool.")
        print(f"Result: {result}")
    except Exception as e:
        print(f"Error: {e}")
    
    print(f"\nHooks fired: {len(hook_log)}")
    for h in hook_log:
        print(f"  - {h}")
    
    return hook_log


async def test_error_handling(hook_log):
    print("\n" + "=" * 60)
    print("TEST 3: Error hook (with bad API key)")
    print("=" * 60)
    
    hook_log.clear()
    print("\nRunning with invalid API key to trigger OnError...")
    config = load_config()
    bad_config = AgentConfig(
        name=config.name,
        model=config.model,
        api_key="invalid-key-trigger-error",
        base_url=config.base_url,
        system_prompt=config.system_prompt,
        temperature=config.temperature,
        timeout_secs=config.timeout_secs,
    )
    bad_agent = Agent.from_config(bad_config)
    
    def log_hook(event, ctx):
        hook_log.append(str(event))
        return "continue"
    
    for event_name in ["BeforeLlmCall", "AfterLlmCall", "BeforeToolCall", 
                       "AfterToolCall", "OnMessage", "OnComplete", "OnError"]:
        bad_agent.register_hook(HookEvent(event_name), log_hook)
    
    try:
        result = await bad_agent.run_simple("test")
        print(f"Result: {result}")
    except Exception as e:
        print(f"Error (expected): {type(e).__name__}")
    
    print(f"\nHooks fired: {len(hook_log)}")
    for h in hook_log:
        print(f"  - {h}")


async def main():
    print("PyBOS Hook & Plugin Full Demo")
    print("=" * 60)
    print(f"\nConfig: {load_config().model}")
    
    agent, hook_log = make_hooks_demo()
    
    await test_run_simple(agent, hook_log)
    await test_react_with_tool(agent, hook_log)
    await test_error_handling(hook_log)
    
    print("\n" + "=" * 60)
    print("SUMMARY")
    print("=" * 60)
    print("""
Hook Events:
  - BeforeLlmCall: Before sending request to LLM
  - AfterLlmCall: After receiving response from LLM
  - BeforeToolCall: Before executing a tool
  - AfterToolCall: After tool execution completes
  - OnMessage: When message is added to conversation
  - OnComplete: When agent completes successfully
  - OnError: When an error occurs

Hook Decisions:
  - 'continue': Proceed normally
  - 'abort': Stop current operation
  - 'error:message': Return error to caller
""")


if __name__ == "__main__":
    asyncio.run(main())